### PR TITLE
WIP Apigee feature installer

### PR DESCRIPTION
**Pull request by bleggett**
Still pending:

- Adding in the pieces from Mike for the Apigee side of things.
- Figuring out a better place to keep @svidhani's apigee gateway function, I am not advocating that we put it here, just did that for convenience.
- No DynamoDB events created yet, pending discussion
- Testing to make sure I have the permissions right

Comments on this basic approach for a standalone feature installer welcome, especially around the IAM/AssumeRole stuff, I _think_ I got that right according to the PolicySimulator.

Note that the Terraform script tries to do the following:

1. Zips up the Apigee gateway function and deploys it to Lambda
2. Creates an IAM Role for the gateway function to execute under, with policies that _only let it invoke functions in the same region, for the same account_
3. Creates a service account IAM User for Apigee to use that ONLY has permissions to execute the gateway function.

### CODE REVIEW

**Comment by suryajak** [comment]
I believe the target branch will be `apigee*`

**Comment by bleggett** [comment]
@suryajak It will but the `apigee-integration` branch is very out of date, to the point where the file structure doesn't match `master`.

Once the current `apigee` branch has been synced with master I will retarget, I won't merge until that sync happens anyway, to avoid getting bitten in the butt by merge conflicts.

File: feature-extensions/apigee/setup.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+
+# TODO Obviously it would be better if we could somehow automatically infer the
+# env prefix that the user used to install Jazz proper, but I don't see a
+# robust way to do that yet, so KISS is a better principal to follow here.
+# Also, there are no programmatic side effects if they happen to enter a
+# different prefix.
+def getEnvPrefix():
+    return raw_input(
+        "Please enter the environment prefix you used for your Jazz install: ")
```
Comment (suryajak) [inline feature-extensions/apigee/setup.py]: To be consistent with the terminology which we use during install, update this "stack prefix" or similar.

File: feature-extensions/apigee/terraform/variables.tf [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# These variables are set by CLI args from the calling script
+variable "region" { type = "string" default = "us-east-1" }
+variable "jazz_aws_accountid" {type = "string"}
+variable "env_prefix" {type = "string"}
```
Comment (suryajak) [inline feature-extensions/apigee/terraform/variables.tf]: Perhaps call this `stack_prefix` here and other places?

File: feature-extensions/apigee/jazz-apigee-proxy/README.md [hunk 3]
```diff
@@ -0,0 +1 @@
+Common Lambda integration for Apigee
```
[the remark this answers is not in the document]
Comment (svidhani) [reply feature-extensions/apigee/jazz-apigee-proxy/README.md]: yes. although this feature is a add-on that admin can select if they want to install (after the jazz installation) apigee being one example here, Ben i believe this should go into the core(jazz repo) as that was my understanding. Did this change and when was this?

File: feature-extensions/apigee/jazz-apigee-proxy/jazz-apigee-proxy.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
+// =========================================================================
```
Comment (svidhani) [inline feature-extensions/apigee/jazz-apigee-proxy/jazz-apigee-proxy.js]: why is this here. This is something already added to core. We need to discuss this

Comment (devsatishm) [reply feature-extensions/apigee/jazz-apigee-proxy/jazz-apigee-proxy.js]: As per the PR description, this is more of an initial take on it. We can discuss and figure out the best location for the code related to these plugins/extensions.

``Figuring out a better place to keep @svidhani's apigee gateway function, I am not advocating that we put it here, just did that for convenience.``

File: feature-extensions/apigee/jazz-apigee-proxy/jazz-apigee-proxy.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =========================================================================
+
+const errorHandlerModule = require("./components/error-handler.js");
+const logger = require("./components/logger.js");
+const AWS = require('aws-sdk');
+
+module.exports.handler = (event, context, callback) => {
```
Comment (devsatishm) [inline feature-extensions/apigee/jazz-apigee-proxy/jazz-apigee-proxy.js]: @bleggett  You might need a sync with the latest version available here: https://github.com/tmobile/jazz/tree/apigee-plugin/jazz_apigee-proxy-aws

File: feature-extensions/apigee/terraform/apigee-gateway-func.tf [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+# Deploy the function with the correct role.
+resource "aws_lambda_function" "jazz-apigee-proxy" {
+  filename         = "${data.archive_file.apigee-gateway-zip.output_path}"
+  function_name    = "${var.env_prefix}-jazz-apigee-proxy"
+  role             = "${aws_iam_role.apigee-lambda-role.arn}"
+  handler          = "jazz-apigee-proxy.handler"
+  source_code_hash = "${data.archive_file.apigee-gateway-zip.output_base64sha256}"
+  runtime          = "nodejs8.10"
```
[the remark this answers is not in the document]
Comment (bleggett) [reply feature-extensions/apigee/terraform/apigee-gateway-func.tf]: This isn't being deployed thru Jazz, it's being deployed by Terraform directly to Lambda, so the fact that it's using the node8 runtime shouldn't affect anything.

Might as well start the move anyway, node10 is about to hit LTS, 8 is what everyone should be on at this point.

File: feature-extensions/apigee/jazz-apigee-proxy/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (bleggett) [reply feature-extensions/apigee/jazz-apigee-proxy/README.md]: I don't disagree at all that the function code should be in `jazz/core`, but since I needed to zip and deploy it for the time being it was simpler to include it in a subdirectory as a temporary measure as I mentioned in the PR comment (also I wanted to prompt this very discussion)

I can add logic that does a shallow Git clone of the main `jazz` repo and zips this one function we care about from the `core` directory.

Should it go into `jazz/core/<fn-name>` or `jazz/core/extension/<fn-name>` so it can be made explicit that this func is an optional part of the system?

Whichever you guys prefer is what I will go with.

File: feature-extensions/apigee/jazz-apigee-proxy/jazz-apigee-proxy.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+      //error thrown when a required parameter is missing
+      if (err.code === 'MissingRequiredParameter') {
+        callback(JSON.stringify(errorHandler.throwMissingParamsError(err.message)));
+      } else {
+        callback(JSON.stringify(errorHandler.throwCommonLambdaError(err.message)));
+      }
+    } else {
+      const responseObj = JSON.parse(data.Payload);
+      if (!responseObj) {
+        logger.error(`No payload response recieved.`);
```
[the remark this answers is not in the document]
Comment (bleggett) [reply feature-extensions/apigee/jazz-apigee-proxy/jazz-apigee-proxy.js]: I copied everything under `jazz-apigee-proxy` from this PR: tmobile/jazz#438, and as discussed I will remove this code from here and pull it from the main jazz repo instead, so @svidhani can you address these?

File: feature-extensions/apigee/jazz-apigee-proxy/package.json [hunk 8]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "jazz_deployments-event-handler",
```
[the remark this answers is not in the document]
Comment (bleggett) [reply feature-extensions/apigee/jazz-apigee-proxy/package.json]: Ditto

File: feature-extensions/apigee/jazz-apigee-proxy/package.json [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  "version": "0.0.1",
+  "description": "Common lambda interator for Apigee",
```
[the remark this answers is not in the document]
Comment (bleggett) [reply feature-extensions/apigee/jazz-apigee-proxy/package.json]: Ditto

File: feature-extensions/apigee/jazz-apigee-proxy/test/test.js [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+  afterEach(() => {
+    event = {};
+    context = {};
+  });
+
+  it("return error message when there is no payload response", () => {
+
+    index.handler(event, context, (err, res) => {
+      expect(err).to.eq("No payload response recieved.");
```
[the remark this answers is not in the document]
Comment (bleggett) [reply feature-extensions/apigee/jazz-apigee-proxy/test/test.js]: Ditto

File: feature-extensions/apigee/jazz-apigee-proxy/test/test.js [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+    });
+  });
+
+  it("return error message when common lamba fails", () => {
```
[the remark this answers is not in the document]
Comment (bleggett) [reply feature-extensions/apigee/jazz-apigee-proxy/test/test.js]: Ditto

File: feature-extensions/apigee/setup.py (hunk 1 above)
Comment (bleggett) [reply feature-extensions/apigee/setup.py]: In our main installer we call the variable `envPrefix`, so I went with that name.

Should it be `stackPrefix` instead?

File: feature-extensions/apigee/terraform/variables.tf (hunk 2 above)
Comment (bleggett) [reply feature-extensions/apigee/terraform/variables.tf]: was attempting to match the existing installer variable name, but agree.

File: feature-extensions/apigee/jazz-apigee-proxy/jazz-apigee-proxy.js (hunk 5 above)
Comment (bleggett) [reply feature-extensions/apigee/jazz-apigee-proxy/jazz-apigee-proxy.js]: As discussed above will remove the code from here and `git clone` as part of the feature extension installer.

File: feature-extensions/apigee/terraform/apigee-gateway-func.tf (hunk 6 above)
[the remark this answers is not in the document]
Comment (suryajak) [reply feature-extensions/apigee/terraform/apigee-gateway-func.tf]: Regarding your first comment, is there a reason we aren't doing this through Jazz? In our process Jazz Admin is able manage all core services (make changes, monitor and all other things just like user services) through their dashboard. Since this will be an exception to the process, I was wondering the rationale behind this.

For second comment, I agree we should move to the latest and greatest at the earliest for all the right reasons. My comment was mostly about how Jazz would deploy it. For the migration, we have almost completed moving off of 4.3, so I believe we are in a good place to start moving to 8.1 for sure..

File: feature-extensions/apigee/jazz-apigee-proxy/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (suryajak) [reply feature-extensions/apigee/jazz-apigee-proxy/README.md]: What is the current plan to enable `apigee`? Is this part of the installation or a post-installation step? I would imagine we will have the first option f(via install flags) while having some process for the second option (post-installation step). 

As part of the installation, `jazz` is already being cloned, hence it makes sense to me that we keep this in `jazz`. 

Regarding the right place in `jazz` - `jazz/core/extension/` makes sense. I might even try to pull this one level up to `jazz/extension`.

File: feature-extensions/apigee/setup.py (hunk 1 above)
Comment (suryajak) [reply feature-extensions/apigee/setup.py]: The installer text presented to the user says something like `stack prefix`. I would retain that text phrasing to keep it consistent. We can choose a different one, but I wouldn't use environment prefix as it has a certain connotation within Jazz so this can get confusing.  

Doing a quick search in installer, I do see that we were being inconsistent in our variable name for this (envPrefix, stack_prefix, stackPrefix). Let us agree on one and then try to make it consistent as we touch the various pieces of code.

File: feature-extensions/apigee/jazz-apigee-proxy/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (svidhani) [reply feature-extensions/apigee/jazz-apigee-proxy/README.md]: i like the idea to have this part of jazz/core/extension. We need to discuss this offline

File: feature-extensions/apigee/jazz-apigee-proxy/jazz-apigee-proxy.js (hunk 4 above)
Comment (svidhani) [reply feature-extensions/apigee/jazz-apigee-proxy/jazz-apigee-proxy.js]: makes sense. we should have a discussion on this

File: feature-extensions/apigee/terraform/apigee-gateway-func.tf (hunk 6 above)
[the remark this answers is not in the document]
Comment (svidhani) [reply feature-extensions/apigee/terraform/apigee-gateway-func.tf]: @suryajak how are we ready to move to 8.1 when lot of our core services are running on 4.3 and have not been upgraded yet. am i missing anything?

[the remark this answers is not in the document]
Comment (bleggett) [reply feature-extensions/apigee/terraform/apigee-gateway-func.tf]: We might need to discuss this more then. My understanding from the last chat was that this function would be optionally deployed into an existing Jazz install after the fact using a standalone installer, and that Jazz admin would merely know if this Jazz installation had Apigee support enabled/installed or not.

If the Apigee gateway function must be managed by Jazz Admin in some way, my understanding of the problem is out of whack.

File: feature-extensions/apigee/jazz-apigee-proxy/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (bleggett) [reply feature-extensions/apigee/jazz-apigee-proxy/README.md]: @suryajak Based on the discussion we had, as an initial step it will be a standalone installer, that is optionally run as a post-installation step. It can later be integrated into the main installer as a flag.

**Comment by bleggett** [comment]
Jacked this branch up by trying some local git stuff, had to delete my fork, but will reopen another PR.
